src/BrainPass/BrainPass.sol

```diff
@@ -0,0 +1,284 @@
+
+        emit NewPassAdded(passId, name, maxTokens, pricePerDay);
+    }
+
+    /// @notice Mint and NFT of a particular passtype
+    /// @param passId The id of the passtype to mint
+    function mintNFT(
+        uint256 passId,
+        uint256 startTimestamp,
+        uint256 endTimestamp
+    ) public {
+        require(
+            addressToPassId[msg.sender][passId] != true,
+            "Max NFTs per address reached"
+        );
+
+        PassType storage passType = passTypes[passId];
+
+        require(passType.maxTokens != 0, "Pass type not found");
+
+        require(
+            passType.lastTokenIdMinted.add(1) <= passType.maxTokens,
+            "Max supply reached"
+        );
+
+        uint256 price = calculatePrice(passId, startTimestamp, endTimestamp);
+
+        if (IERC20(IqToken).balanceOf(msg.sender) < price)
+            revert UserBalanceNotEnough();
+
+        bool success = IERC20(IqToken).transferFrom(msg.sender, owner, 1e18);
+        if (!success) revert MintingPaymentFailed();
+
+        uint256 tokenId = passType.lastTokenIdMinted + 1;
+        setBaseURI(passType.tokenURI);
+        _safeMint(msg.sender, tokenId);
+
+        addressToPassId[msg.sender][passId] = true;
+
+        UserPassItem memory purchase = UserPassItem(
+            tokenId,
+            passId,
+            startTimestamp,
+            endTimestamp
+        );
+        addressToNFTPass[msg.sender][tokenId] = purchase;
+        passType.lastTokenIdMinted = tokenId;
+
+        emit BrainPassBought(
+            msg.sender,
+            tokenId,
+            passId,
+            startTimestamp,
+            endTimestamp
+        );
+    }
+
+    /// @notice Calculate the price of an Nft
+    /// @param startTimestamp and endTimestamp are used to calc the price to be paid
+    function calculatePrice(
+        uint256 passId,
+        uint256 startTimestamp,
+        uint256 endTimestamp
+    ) public view returns (uint256) {
+        PassType memory passType = passTypes[passId];
+        //duration in days
+        uint256 duration = endTimestamp.sub(startTimestamp);
+        uint256 totalPrice = duration.mul(passType.pricePerDay);
+        if (passType.discount > 0) {
+            uint256 discountAmount = totalPrice.mul(passType.discount).div(100);
+            totalPrice = totalPrice.sub(discountAmount);
+        }
+
+        return totalPrice;
+    }
+
+    /// @notice Increase the time to hold a PassNft
+    /// @param tokenId The Id of the NFT whose time is to be increased
+    function increasePassTime(
+        uint256 tokenId,
+        uint256 passIdNum,
+        uint newStartTime,
+        uint256 newEndTime
+    ) public {
+        require(
+            msg.sender == ownerOf(tokenId),
+            "You cannot increase the time for an NFT you don't own"
+        );
+
+        UserPassItem storage pass = addressToNFTPass[ownerOf(tokenId)][tokenId];
+        uint256 price = calculatePrice(passIdNum, newStartTime, newEndTime);
```

Comment:
bug: malicious user can use a different passId to increase time in a cheaper passId. you dont need passIdNum as param since you can get it from the user pass

Reply:
Yes Ill improve the IncreaseTime Func